### ISSUE TITLE
Suggest same simplifications for (&) as for ($)

### ISSUE THREAD

**Issue by 3noch** [ndmitchell/hlint#399]
Many suggestions work for simplifying expressions with `($)` but not with `(&)`. For example:

HLint correctly suggests that the following be eta reduced:

```haskell
f x = g $ x
```

But it doesn't suggest that for the following:

```haskell
f x = x & g
```

**Comment by ndmitchell** [comment]
When HLint was written & wasn't a function! Sounds reasonable - anywhere else you think & needs special treatment?

**Comment by 3noch** [comment]
That's true! Because of that, it's probably worth considering that `(&)` is sometimes used for other purposes since it was added later in the game: http://hoogle.haskell.org/?hoogle=%28%26%29&scope=set%3Astackage

You would know best how to address that concern.

The only other suggestions I can think of would relate to reducing parentheses, in similar scenarios as `($)`.

**Comment by ndmitchell** [comment]
It's also harder because it's not in the Prelude. Most of the magic in HLint applies only to `.` and `$`, and treats them as though they always appear unqualified and are always in scope (basically because they are).

I'm also wondering if people write `f x = x & y` deliberately? I suspect they might - it's typically done with lens stuff and seeing `(& y)` there would look very weird.

**Comment by 3noch** [comment]
Well you'd just see `y`, not `(& y)`. I came across this for that reason. I was using lenses and realized that I could eta-reduce but hlint hadn't said anything.

However, the other difficulties may be enough to forego this feature. If it's easy to add and also easy to disable then it might be worth it. Otherwise probably not.

**Comment by ndmitchell** [comment]
OK, that makes sense - I think eta expanding `&` in situations where the `&` can disappear makes sense.

**Comment by 3noch** [comment]
By that do you mean that you believe it unlikely that folks will want to reduce `x & g` to just `g`?

**Comment by ndmitchell** [comment]
I believe that's the only place people will want to do it - when the & can be removed.

**Comment by 3noch** [comment]
Yeah, just like with `$`. And if it's a single expression as the definition of a function, that's when the suggestion comes into play for either operator.

**Comment by 3noch** [comment]
Thanks!
